main.cpp

moe-salih opened the pull request:
new sieve implementation maybe.
i don't know if this is really new implementation i just got this idea and it gave me good result on my pc when I run it. 👍hope you like it

## Description
<!--
Add your description yere.
-->

## Contributing requirements
<!--
Make sure your PR conforms to the requirements set out in CONTRIBUTING.md:
-->

<!--
When ticking below boxes, please don't leave spaces between the 'x' and the square brackets, as that breaks the checkbox rendering in the PRs.
Right: [x]
Wrong: [x ]
-->
* [ ] I read the contribution guidelines in CONTRIBUTING.md.
* [ ] I placed my solution in the correct solution folder.
* [ ] I added a README.md with the right badge(s).
* [ ] I added a Dockerfile that builds and runs my solution.
* [ ] I selected `drag-race` as the target branch.
* [ ] All code herein is licensed compatible with BSD-3.

rhbvkleef commented:
You ABSOLUTE massochist :)

dasper commented:
Nice. For the fun of it I checked out your implementation and compared it to using just an array of bools in PHP7.4 on my Ryzen 3900X.

- Windows
bits `Passes: 97, Time: 10.029282, Avg: 0.103395, Limit: 1000000, Count: 78498, Valid: 1`
bools `Passes: 128, Time: 10.056836, Avg: 0.078569, Limit: 1000000, Count: 78498, Valid: 1`
- PopOS
bits `Passes: 149, Time: 10.029282, Avg: 0.103395, Limit: 1000000, Count: 78498, Valid: 1`
bools `Passes: 196, Time: 10.056836, Avg: 0.078569, Limit: 1000000, Count: 78498, Valid: 1`

AJenbo commented:
While many bad things have been said about PHP it's actually the case that it's one of the fastest interpreted languages since (since PHP 7).

I noticed that like my self you are a fan of type checking. Note that this is done at run time in PHP and so you could gain about 8% performance by stripping it. You could use old-style PHPDoc in its place, and as you mention it would then also be compatible with PHP 5

With type checking:
php7.0 `Passes: 125, Time: 10.033035, Avg: 0.080264, Limit: 1000000, Count: 78498, Valid: 1`
php7.1 `Passes: 133, Time: 10.024397, Avg: 0.075371, Limit: 1000000, Count: 78498, Valid: 1`
php7.2 `Passes: 142, Time: 10.040123, Avg: 0.070705, Limit: 1000000, Count: 78498, Valid: 1`
php7.3 `Passes: 157, Time: 10.049980, Avg: 0.064013, Limit: 1000000, Count: 78498, Valid: 1`
php7.4 `Passes: 153, Time: 10.001033, Avg: 0.065366, Limit: 1000000, Count: 78498, Valid: 1`
php8.0 `Passes: 150, Time: 10.043180, Avg: 0.066955, Limit: 1000000, Count: 78498, Valid: 1`

No type checking:
php5.6 `Passes: 57, Time: 10.150949, Avg: 0.178087, Limit: 1000000, Count: 78498, Valid: 1`
php7.0 `Passes: 135, Time: 10.055309, Avg: 0.074484, Limit: 1000000, Count: 78498, Valid: 1`
php7.1 `Passes: 141, Time: 10.069466, Avg: 0.071415, Limit: 1000000, Count: 78498, Valid: 1`
php7.2 `Passes: 163, Time: 10.005253, Avg: 0.061382, Limit: 1000000, Count: 78498, Valid: 1`
php7.3 `Passes: 163, Time: 10.032652, Avg: 0.061550, Limit: 1000000, Count: 78498, Valid: 1`
php7.4 `Passes: 161, Time: 10.041914, Avg: 0.062372, Limit: 1000000, Count: 78498, Valid: 1`
php8.0 `Passes: 159, Time: 10.058095, Avg: 0.063258, Limit: 1000000, Count: 78498, Valid: 1`

Lastly PHP JIT is not enabled when called from the CLI, to do that you have to tell it explicitly to use the JIT:

```
php8.0 -dopcache.enable_cli=1 -dopcache.jit_buffer_size=100M PrimePHP.php
Passes: 363, Time: 10.008409, Avg: 0.027571, Limit: 1000000, Count: 78498, Valid: 1
```

For comparison:
Python 3.8 `Passes: 63, Time: 10.132384637996438, Avg: 0.16083150219041964, Limit: 1000000, Count: 78498, Valid: True`
GCC 10.2.0 `Passes: 9723, Time: 5.000000, Avg: 0.000514, Limit: 1000000, Count1: 78498, Count2: 78498, Valid: 1`

(Ryzen 3700X)

rbergen commented:
@nyirocsaba We are happy to work towards including your solution in the drag-race; we don't yet have a PHP implementation in the repo that uses bit-level prime candidate storage. However, to be able to do so, it would need to comply with the [contributing guidelines](https://github.com/PlummersSoftwareLLC/Primes/blob/drag-race/CONTRIBUTING.md). That also means that the PR needs to be opened on the `drag-race` branch of this repo, not `original` (formerly `main`).

rbergen commented:
PHP improvements have been merged with #307. This PR can be reopened on drag-race if so desired.

moe-salih committed:
main.cpp

new sieve implementation maybe.
i don't know if this is really new implementation i just got this idea and it gave me good result on my pc when I run it. 👍hope you like it
